### PR TITLE
Run Tests during CI

### PR DESCRIPTION
**Pull request by sharath-cliqz**

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -8,11 +8,31 @@ node('us-east-1 && ubuntu && docker && !gpu') {
         checkout scm
         def delim = '"'
         hash = sh(returnStdout: true, script: """curl https://api.github.com/repos/cliqz-oss/cliqz-android/branches/upstream-releases | grep -m 1 sha | cut -d '${delim}' -f 4""").trim()
+        withCredentials([file(credentialsId: 'ceb2d5e9-fc88-418f-aa65-ce0e0d2a7ea1', variable: 'CLIQZ_CI_SSH_KEY')]) {
```
Comment (chrmod) [inline Jenkinsfile]: would be nice to wrap it in a helper, also to clean it up (with try/finally) after block is executed

File: Jenkinsfile [hunk 2]
```diff
@@ -8,11 +8,31 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
+                git clone -b version2.0 --single-branch --depth=1 git@github.com:cliqz/autobots.git
+            '''
+        }
     }
     def dockerTag = "${hash}"
     def baseImageName = "browser-f/android:${dockerTag}"
-    docker.image("141047255820.dkr.ecr.us-east-1.amazonaws.com/${baseImageName}").inside {
-        try {
+    def instance_id = sh(returnStdout: true, script: '''
+            aws ec2 run-instances --image-id ami-13050368 --count 1 --instance-type t2.medium --key-name android_ci_genymotion --security-group-ids sg-5bbf173f --subnet-id subnet-341ff61f  --region=us-east-1 --query "Instances[].InstanceId" --output text
```
Comment (chrmod) [inline Jenkinsfile]: how about splitting this long string into multiple lines?

File: Jenkinsfile [hunk 3]
```diff
@@ -8,11 +8,31 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
+        ''').trim()
+    def ip = sh(returnStdout: true, script: """
+            aws ec2 describe-instances --instance-ids $instance_id  --region=us-east-1 --query 'Reservations[].Instances[].PrivateIpAddress' --output text
+        """).trim()
+    sh "aws ec2 create-tags --resources $instance_id --region=us-east-1 --tag Key=Name,Value='Appium-Genymotion'"
+    def apk = "mozilla-release/objdir-frontend-android/dist/fennec-59.0.2.en-US.android-i386.apk"
+    try {
+        docker.image("141047255820.dkr.ecr.us-east-1.amazonaws.com/${baseImageName}").inside {
```
Comment (chrmod) [inline Jenkinsfile]: you can use `withRegistry` helper

File: Jenkinsfile [hunk 4]
```diff
@@ -23,17 +43,80 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
-            stage('Upload APK') {
-                archiveArtifacts allowEmptyArchive: true, artifacts: 'mozilla-release/objdir-frontend-android/dist/*.apk'
+        }
+        withEnv(["INSTANCE_ID=${instance_id}"]){
+            timeout(5){
+                stage('Genymotion Status'){
+                    def status = sh(returnStdout: true, script: """
+                            aws ec2 describe-instance-status --region=us-east-1 --instance-id $INSTANCE_ID --query 'InstanceStatuses[].InstanceStatus[].Details[].Status' --output text
+                        """).trim()
+                    while (status != 'passed') {
```
Comment (chrmod) [inline Jenkinsfile]: is there no better way?

File: Jenkinsfile [hunk 5]
```diff
@@ -23,17 +43,80 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
         }
-        finally {
-            stage('Clean Up') {
-                sh '''#!/bin/bash -l
-                    rm -f mozilla-release/mozconfig
-                    rm -rf mozilla-release/objdir-frontend-android
-                '''
+        docker.image("141047255820.dkr.ecr.us-east-1.amazonaws.com/${baseImageName}").inside {
+            withEnv(["IP=${ip}", "APP=${apk}", "platformName=android", "deviceName=127.0.0.1:5556"]){
+                withCredentials([file(credentialsId: 'e87da5bf-eb00-4d85-8ef9-93680abaa266', variable: 'FILE' )]){
```
Comment (chrmod) [inline Jenkinsfile]: what is this FILE exactly?

File: Jenkinsfile [hunk 6]
```diff
@@ -23,17 +43,80 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
             }
         }
-        finally {
-            stage('Clean Up') {
-                sh '''#!/bin/bash -l
-                    rm -f mozilla-release/mozconfig
-                    rm -rf mozilla-release/objdir-frontend-android
-                '''
+        docker.image("141047255820.dkr.ecr.us-east-1.amazonaws.com/${baseImageName}").inside {
+            withEnv(["IP=${ip}", "APP=${apk}", "platformName=android", "deviceName=127.0.0.1:5556"]){
```
Comment (chrmod) [inline Jenkinsfile]: please split into multiple lines

File: Jenkinsfile [hunk 7]
```diff
@@ -23,17 +43,80 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
+                                chmod 400 $FILE
+                                ssh -v -o StrictHostKeyChecking=no -i $FILE root@$IP "setprop persist.sys.usb.config adb"
+                                ssh -v -o StrictHostKeyChecking=no -i $FILE -NL 5556:127.0.0.1:5555 root@$IP &
+                                ~/android_home/platform-tools/adb connect 127.0.0.1:5556
+                                ~/android_home/platform-tools/adb wait-for-device
+                                appium &
+                                sleep 10
+                                export app=$PWD/$APP
+                                cd autobots
+                                chmod 0755 requirements.txt
```
Comment (chrmod) [inline Jenkinsfile]: why this is needed?

File: Jenkinsfile [hunk 8]
```diff
@@ -23,17 +43,80 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
+                                virtualenv ~/venv
+                                source ~/venv/bin/activate
+                                pip install -r requirements.txt
+                                python testRunner.py || true
+                           '''
+                       }
+                    }
+                }
+            }
+            try{
```
Comment (chrmod) [inline Jenkinsfile]: space after try is missing :-)

File: Jenkinsfile [hunk 9]
```diff
@@ -23,17 +43,80 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
+                stage('Upload APK') {
+                    archiveArtifacts allowEmptyArchive: true, artifacts: 'mozilla-release/objdir-frontend-android/dist/*.apk'
+                    archiveArtifacts allowEmptyArchive: true, artifacts: 'autobots/*.log'
+                    junit "autobots/test-reports/*.xml"
```
Comment (chrmod) [inline Jenkinsfile]: no need for double quotes

File: Jenkinsfile [hunk 10]
```diff
@@ -23,17 +43,80 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
+                                pip install -r requirements.txt
+                                python testRunner.py || true
+                           '''
+                       }
+                    }
+                }
+            }
+            try{
+                stage('Upload APK') {
+                    archiveArtifacts allowEmptyArchive: true, artifacts: 'mozilla-release/objdir-frontend-android/dist/*.apk'
```
Comment (chrmod) [inline Jenkinsfile]: why do we archive apk? archives are stored on Jenkins master, maybe we should upload it to S3 instead?

File: Jenkinsfile [hunk 11]
```diff
@@ -23,17 +43,80 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
+                    archiveArtifacts allowEmptyArchive: true, artifacts: 'autobots/*.log'
+                    junit "autobots/test-reports/*.xml"
+                    zip archive: true, dir: 'autobots/screenshots', glob: '', zipFile: 'autobots/screenshots.zip'
+                }
             }
+            catch(e) {
+                print e
+                // no screenshots, no problem
```
Comment (chrmod) [inline Jenkinsfile]: but this catch covers errors not only from screenshots

File: Jenkinsfile [hunk 12]
```diff
@@ -23,17 +43,80 @@ node('us-east-1 && ubuntu && docker && !gpu') {
...
+        print e
+    }
+    finally {
+        sh """#!/bin/bash -l
+            set -x
+            set -e
+            aws ec2 terminate-instances --instance-ids ${instance_id} --region=us-east-1
+        """
+        stage('Clean Up') {
+            sh '''#!/bin/bash -l
```
Comment (chrmod) [inline Jenkinsfile]: why we need `-l` here?

Comment (sharath-cliqz) [reply Jenkinsfile]: Removed -l

File: Jenkinsfile (hunk 11 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Yes. Removed the comment.

File: Jenkinsfile (hunk 9 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Changed it to single quote.

File: Jenkinsfile (hunk 10 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: It is to make sure that if the test's fail, it is easier to debug with the apk, and not pull the same commit and rebuilding it.
We are also storing only the latest 25 builds.

File: Jenkinsfile (hunk 8 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Added Space :D

File: Jenkinsfile (hunk 7 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: I remember, it not working/installing using PIP properly before adding this.
Removing it for now.

File: Jenkinsfile (hunk 5 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Credentials To connect to the Genymotion Instance.

File: Jenkinsfile (hunk 6 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 4 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Other methods are flaky, and has a failure rate of 30-50%. This way, it is working 100%.

File: Jenkinsfile (hunk 3 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Yes. Done.

File: Jenkinsfile (hunk 2 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Done.

File: Jenkinsfile [hunk 13]
```diff
@@ -2,38 +2,194 @@
...
         }
     }
 }
+
+@NonCPS
+def withGenymotion(Closure body) {
+    try {
+        def instance_id = sh(returnStdout: true, script: '''
+            aws ec2 run-instances \
+            --image-id ami-13050368 \
```
Comment (chrmod) [inline Jenkinsfile]: I would indent subsequent arguments a bit

File: Jenkinsfile [hunk 14]
```diff
@@ -2,38 +2,194 @@
...
+            --key-name android_ci_genymotion \
+            --security-group-ids sg-5bbf173f \
+            --subnet-id subnet-341ff61f  \
+            --region=us-east-1 \
+            --query "Instances[].InstanceId" \
+            --output text
+            ''').trim()
+        def ip = sh(returnStdout: true, script: """
+            aws ec2 describe-instances \
+            --instance-ids $instance_id \
```
Comment (chrmod) [inline Jenkinsfile]: I like to add curly braces around groovy variables so they are easier to distinguish from environment variables

File: Jenkinsfile [hunk 15]
```diff
@@ -2,38 +2,194 @@
...
+            --output text
+            """).trim()
+        sh "aws ec2 create-tags \
+        --resources $instance_id \
+        --region=us-east-1 \
+        --tag Key=Name,Value='Appium-Genymotion'"
+        withEnv([
+            "INSTANCE_ID=${instance_id}",
+            "IP=${ip}"
+            ]){
```
Comment (chrmod) [inline Jenkinsfile]: indent out?

File: Jenkinsfile [hunk 16]
```diff
@@ -2,38 +2,194 @@
...
+                docker.image("${baseImageName}").inside {
+                    withEnv([
+                        "APP=${apk}",
+                        "platformName=android",
+                        "deviceName=127.0.0.1:5556",
+                        "MODULE=testSmoke",
+                        "TEST=SmokeTest",
+                        "appPackage=org.mozilla.fennec_",
+                        "appActivity=org.mozilla.gecko.LauncherActivity"
+                        ]) {
```
Comment (chrmod) [inline Jenkinsfile]: indent out?

File: Jenkinsfile [hunk 17]
```diff
@@ -2,38 +2,194 @@
...
-        checkout scm
-        def delim = '"'
-        hash = sh(returnStdout: true, script: """curl https://api.github.com/repos/cliqz-oss/cliqz-android/branches/upstream-releases | grep -m 1 sha | cut -d '${delim}' -f 4""").trim()
-    }
-    def dockerTag = "${hash}"
-    def baseImageName = "browser-f/android:${dockerTag}"
-    docker.image("141047255820.dkr.ecr.us-east-1.amazonaws.com/${baseImageName}").inside {
+    def dockerTag = ""
+    def apk = ""
+    // def instance_id = sh(returnStdout: true, script: '''
```
Comment (chrmod) [inline Jenkinsfile]: can those comments be removed?

File: Jenkinsfile [hunk 18]
```diff
@@ -2,38 +2,194 @@
...
+                            }
+                        }
+                        stage('Run Tests') {
+                            timeout(60) {
+                                sh'''#!/bin/bash -l
+                                    set -x
+                                    set -e
+                                    appium &
+                                    sleep 10
+                                    export app=$PWD/mozilla-release/objdir-frontend-android/dist/$APP
```
Comment (chrmod) [inline Jenkinsfile]: `withEnv`?

Comment (sharath-cliqz) [reply Jenkinsfile]: The Path changes with every PR. So had to export it from the PR Directory.

File: Jenkinsfile (hunk 17 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Trying them out to check if it works. Removed them.

File: Jenkinsfile (hunk 16 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 15 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Done.

File: Jenkinsfile (hunk 14 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Yea.. I missed a couple of them. Added now. Done.

File: Jenkinsfile (hunk 13 above)
Comment (sharath-cliqz) [reply Jenkinsfile]: Done.

**Comment by spacifici** [comment]
@chrmod should I merge this?

File: Jenkinsfile [hunk 19]
```diff
@@ -2,38 +2,201 @@
...
-                    set -e
-                    set -x
-                    cd mozilla-release
-                    mv ../mozconfigs/cliqz.mozconfig mozconfig
-                    ./mach build
-                    ./mach package
-                '''
+    def dockerTag = ""
+    def apk = ""
+    try {
```
Comment (chrmod) [inline Jenkinsfile]: why we try/catch here? shouldn't it be handled in `withGenymotion`

File: Jenkinsfile [hunk 20]
```diff
@@ -2,38 +2,201 @@
...
+    } finally {
+        sh """#!/bin/bash -l
+            set -x
+            set -e
+            aws ec2 terminate-instances --instance-ids ${instance_id} --region=${region}
+        """
     }
 }
+
+@NonCPS
```
Comment (chrmod) [inline Jenkinsfile]: it does not look like it is needed

**Pull request merged**